base_layer/wallet/src/output_manager_service/output_manager_service.rs

```diff
@@ -0,0 +1,413 @@
+            .get_mut(&tx_id)
+            .ok_or(OutputManagerError::PendingTransactionNotFound)?;
+
+        if pending_transaction.outputs_to_be_received.len() != 1 ||
+            pending_transaction.outputs_to_be_received[0]
+                .as_transaction_input(&COMMITMENT_FACTORY, OutputFeatures::empty())
+                .commitment !=
+                received_output.commitment
+        {
+            return Err(OutputManagerError::IncompleteTransaction);
+        }
+
+        self.unspent_outputs
+            .append(&mut pending_transaction.outputs_to_be_received);
+        let _ = self.pending_transactions.remove(&tx_id);
+        Ok(())
+    }
+
+    /// Prepare a Sender Transaction Protocol for the amount and fee_per_gram specified. If required a change output
+    /// will be produced.
+    pub fn prepare_transaction_to_send(
+        &mut self,
+        amount: MicroTari,
+        fee_per_gram: MicroTari,
+        lock_height: Option<u64>,
+    ) -> Result<SenderTransactionProtocol, OutputManagerError>
+    {
+        let mut rng = TransactionRng::new().unwrap();
+        let outputs = self.select_outputs(amount, fee_per_gram, UTXOSelectionStrategy::Smallest)?;
+        let total = outputs.iter().fold(MicroTari::from(0), |acc, x| acc + x.value);
+
+        let offset = PrivateKey::random(&mut rng);
+        let nonce = PrivateKey::random(&mut rng);
+
+        let mut builder = SenderTransactionProtocol::builder(1);
+        builder
+            .with_lock_height(lock_height.unwrap_or(0))
+            .with_fee_per_gram(fee_per_gram)
+            .with_offset(offset.clone())
+            .with_private_nonce(nonce.clone())
+            .with_amount(0, amount);
+
+        for uo in outputs.iter() {
+            builder.with_input(
+                uo.as_transaction_input(&COMMITMENT_FACTORY, OutputFeatures::empty()),
+                uo.clone(),
+            );
+        }
+
+        let fee_without_change = Fee::calculate(fee_per_gram, outputs.len(), 1);
+        let mut change_key: Option<PrivateKey> = None;
+        // If the input values > the amount to be sent + fees_without_change then we will need to include a change
+        // output
+        if total > amount + fee_without_change {
+            let mut km = acquire_lock!(self.key_manager);
+            let key = km.next_key()?.k;
+            change_key = Some(key.clone());
+            builder.with_change_secret(key);
+        }
+
+        let stp = builder
+            .build::<HashDigest>(&PROVER, &COMMITMENT_FACTORY)
+            .map_err(|e| OutputManagerError::BuildError(e.message))?;
+
+        // The Transaction Protocol built successfully so we will pull the unspent outputs out of the unspent list and
+        // store them until the transaction times out OR is confirmed
+        let mut outputs_to_be_spent = self.unspent_outputs.clone();
+        outputs_to_be_spent.retain(|uo| outputs.iter().any(|o| uo.spending_key == o.spending_key));
```

Comment:
I suspect something like
`let outputs_to_be_spent = self.unspent_outputs.iter().filter_map(|utxo| outputs.iter().find(|o| uo.spending_key == o.spending_key))`

involves less memory shunting

Reply:
Figured out how to turn on the experimental drain_filter() feature for vecs and used that.